File: bench/wordfreq/wordfreq.go
// Go peer of bench/wordfreq.vader. The corpus is a []byte, and each token is
// turned into a string with `string(corpus[start:end])` — the conversion Go
// tokenizers write, which allocates and copies the bytes before the map hashes
// them.

package main

import "fmt"

const (
	vocabularySize     = 4096
	tokenCount         = 300_000
	wordLength         = 6
	scrambleMultiplier = 2654435761
	vocabularyStride   = 2731
	vocabularySpace    = 308_915_776
)

func appendWord(out []byte, index int32) []byte {
	scrambled := int64(index) * scrambleMultiplier % vocabularySpace
	for position := 0; position < wordLength; position++ {
		out = append(out, byte('a'+scrambled%26))
		scrambled = scrambled / 26
	}
	return out
}

// Rendered ONCE : the corpus repeats each word 73 times and the checksum reads
// every one back.
func buildVocabulary() [][]byte {
	out := make([][]byte, 0, vocabularySize)
	for index := 0; index < vocabularySize; index++ {
		out = append(out, appendWord(make([]byte, 0, wordLength), int32(index)))
	}
	return out
}

func main() {
	vocabulary := buildVocabulary()
	corpus := make([]byte, 0, tokenCount*(wordLength+1))
	for tokenIndex := 0; tokenIndex < tokenCount; tokenIndex++ {
		wordIndex := int64(tokenIndex) * vocabularyStride % vocabularySize
		corpus = append(corpus, vocabulary[wordIndex]...)
		corpus = append(corpus, ' ')
	}

	counts := make(map[string]int32)
	for tokenIndex := 0; tokenIndex < tokenCount; tokenIndex++ {
		start := tokenIndex * (wordLength + 1)
		token := string(corpus[start : start+wordLength])
		counts[token] = counts[token] + 1
	}

	var checksum int64
	for index := 0; index < vocabularySize; index++ {
		if seen, ok := counts[string(vocabulary[index])]; ok {
			checksum += int64(seen) * int64(index+1)
		}
	}
	fmt.Printf("wordfreq vocabulary=%d tokens=%d distinct=%d checksum=%d\n",
		vocabularySize, tokenCount, len(counts), checksum)
}

File: bench/wordfreq/wordfreq.java
// Java peer of bench/wordfreq.vader. The corpus is a byte[], and each token is
// turned into a String with the offset/length constructor — the conversion a
// Java tokenizer writes, which copies the bytes before the map hashes them.

import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Map;

static final int VOCABULARY_SIZE = 4096;
static final int TOKEN_COUNT = 300_000;
static final int WORD_LENGTH = 6;
static final long SCRAMBLE_MULTIPLIER = 2654435761L;
static final long VOCABULARY_STRIDE = 2731L;
static final long VOCABULARY_SPACE = 308_915_776L;

static int appendWord(byte[] out, int at, int index) {
    long scrambled = index * SCRAMBLE_MULTIPLIER % VOCABULARY_SPACE;
    for (int position = 0; position < WORD_LENGTH; position++) {
        out[at + position] = (byte) ('a' + scrambled % 26);
        scrambled = scrambled / 26;
    }
    return at + WORD_LENGTH;
}

static String vocabularyWord(int index) {
    byte[] out = new byte[WORD_LENGTH];
    appendWord(out, 0, index);
    return new String(out, StandardCharsets.US_ASCII);
}

void main() {
    byte[] corpus = new byte[TOKEN_COUNT * (WORD_LENGTH + 1)];
    for (int tokenIndex = 0; tokenIndex < TOKEN_COUNT; tokenIndex++) {
        int wordIndex = (int) (tokenIndex * VOCABULARY_STRIDE % VOCABULARY_SIZE);
        int at = appendWord(corpus, tokenIndex * (WORD_LENGTH + 1), wordIndex);
        corpus[at] = ' ';
    }

    Map<String, Integer> counts = new HashMap<>();
    for (int tokenIndex = 0; tokenIndex < TOKEN_COUNT; tokenIndex++) {
        String token = new String(
            corpus, tokenIndex * (WORD_LENGTH + 1), WORD_LENGTH, StandardCharsets.US_ASCII);
        counts.merge(token, 1, Integer::sum);
    }

    long checksum = 0;
    for (int index = 0; index < VOCABULARY_SIZE; index++) {
        Integer seen = counts.get(vocabularyWord(index));
        if (seen != null) checksum += (long) seen * (index + 1);
    }
    IO.println("wordfreq vocabulary=%d tokens=%d distinct=%d checksum=%d"
        .formatted(VOCABULARY_SIZE, TOKEN_COUNT, counts.size(), checksum));
}
